nodebuilder/header/data_commitment.go

```diff
@@ -0,0 +1,242 @@
+
+// validateDataCommitmentRange runs basic checks on the asc sorted list of
+// heights that will be used subsequently in generating data commitments over
+// the defined set of heights.
+func (s *Service) validateDataCommitmentRange(ctx context.Context, start, end uint64) error {
+	if start == 0 {
+		return fmt.Errorf("the start block is 0")
+	}
+	if start >= end {
+		return fmt.Errorf("end block is smaller or equal to the start block")
+	}
+	heightsRange := end - start
+	if heightsRange > uint64(dataCommitmentBlocksLimit) {
+		return fmt.Errorf("the query exceeds the limit of allowed blocks %d", dataCommitmentBlocksLimit)
+	}
+
+	currentHeader, err := s.NetworkHead(ctx)
+	if err != nil {
+		return err
+	}
+	// the data commitment range is end exclusive
+	if end > currentHeader.Height()+1 {
+		return fmt.Errorf(
+			"end block %d is higher than current chain height %d",
+			end,
+			currentHeader.Height(),
+		)
+	}
+
+	currentLocalHeader, err := s.LocalHead(ctx)
+	if err != nil {
+		return err
+	}
+	// the data commitment range is end exclusive
+	if end > currentLocalHeader.Height()+1 {
+		return fmt.Errorf(
+			"end block %d is higher than local chain height %d. Wait for the node until it syncs up to %d",
+			end,
+			currentLocalHeader.Height(),
+			end,
+		)
+	}
+	return nil
+}
+
+// hashDataRootTuples hashes a list of blocks data root tuples, i.e., height, data root and square
+// size, then returns their merkle root.
+func hashDataRootTuples(tuples []DataRootTuple) ([]byte, error) {
+	if len(tuples) == 0 {
+		return nil, fmt.Errorf("cannot hash an empty list of data root tuples")
+	}
+	dataRootEncodedTuples := make([][]byte, 0, len(tuples))
+	for _, tuple := range tuples {
+		encodedTuple, err := EncodeDataRootTuple(
+			tuple.height,
+			tuple.dataRoot,
+		)
+		if err != nil {
+			return nil, err
+		}
+		dataRootEncodedTuples = append(dataRootEncodedTuples, encodedTuple)
+	}
+	root := merkle.HashFromByteSlices(dataRootEncodedTuples)
+	return root, nil
+}
+
+// validateDataRootInclusionProofRequest validates the request to generate a data root
+// inclusion proof.
+func (s *Service) validateDataRootInclusionProofRequest(
+	ctx context.Context,
+	height, start, end uint64,
+) error {
+	err := s.validateDataCommitmentRange(ctx, start, end)
+	if err != nil {
+		return err
+	}
+	if height < start || height >= end {
+		return fmt.Errorf(
+			"height %d should be in the end exclusive interval first_block %d last_block %d",
+			height,
+			start,
+			end,
+		)
+	}
+	return nil
+}
+
+// proveDataRootTuples returns the merkle inclusion proof for a height.
+func proveDataRootTuples(tuples []DataRootTuple, height int64) (*merkle.Proof, error) {
+	if len(tuples) == 0 {
+		return nil, fmt.Errorf("cannot prove an empty list of tuples")
+	}
+	if height < 0 {
+		return nil, fmt.Errorf("cannot prove a strictly negative height %d", height)
+	}
+	currentHeight := tuples[0].height - 1
+	for _, tuple := range tuples {
+		if tuple.height != currentHeight+1 {
```

Comment:
this function is being called after `fetchDataRootTuples`, which - if you use `GetRange..` request, guarantees a contiguous ascending range to be returned. 

You do not need to do this validation logic here, function can document that it assumes the tuples to be contiguous and complete.